pkg/csi/service/vanilla/controller.go

```diff
@@ -1137,15 +1139,109 @@ func (c *controller) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshot
+	}
+
+	// Validate arguments
+	volumeID := req.GetSourceVolumeId()
+	if len(volumeID) == 0 {
+		return nil, logger.LogNewErrorCode(log, codes.InvalidArgument,
+			"CreateSnapshot Source Volume ID must be provided")
+	}
+
+	if len(req.Name) == 0 {
+		return nil, logger.LogNewErrorCode(log, codes.InvalidArgument,
+			"Snapshot name must be provided")
+	}
+
+	// Check if volume already exists
+	volumeIds := []cnstypes.CnsVolumeId{{Id: volumeID}}
+	queryFilter := cnstypes.CnsQueryFilter{
+		VolumeIds: volumeIds,
+	}
+	queryResult, err := c.manager.VolumeManager.QueryVolume(ctx, queryFilter)
+	if err != nil {
+		return nil, logger.LogNewErrorCodef(log, codes.Internal,
+			"The volume %q to be snapshotted is not available: %v", volumeID, err)
+	}
+
+	// Get the snapshot size by getting the volume size, which is only valid for full backup use cases
+	var snapshotSize int64
+	if len(queryResult.Volumes) > 0 {
+		snapshotSize = queryResult.Volumes[0].BackingObjectDetails.GetCnsBackingObjectDetails().CapacityInMb
+	} else {
+		return nil, logger.LogNewErrorCodef(log, codes.Internal,
+			"failed to get the snapshot size by querying volume: %q", volumeID)
+	}
+
+	snapshotID, snapshotCreateTime, err := common.CreateSnapshotUtil(ctx, c.manager, volumeID, req.Name)
+	if err != nil {
+		return nil, logger.LogNewErrorCodef(log, codes.Internal,
+			"failed to create snapshot on volume %q: %v", volumeID, err)
+	}
+
+	snapshotCreateTimeInProto := timestamppb.New(*snapshotCreateTime)
+	if err := snapshotCreateTimeInProto.CheckValid(); err != nil {
+		return nil, logger.LogNewErrorCodef(log, codes.Internal,
+			"failed to convert creation timestamp to proto format: %v", err)
+	}
+
+	createSnapshotResponse := &csi.CreateSnapshotResponse{
+		Snapshot: &csi.Snapshot{
+			SizeBytes:      snapshotSize * common.MbInBytes,
+			SnapshotId:     snapshotID,
+			SourceVolumeId: volumeID,
+			CreationTime:   snapshotCreateTimeInProto,
+			ReadyToUse:     true,
+		},
+	}
+
+	log.Infof("CreateSnapshot succeeded for snapshot %s "+
+		"on volume %s size %d Time proto %d Timestamp %+v Response: %+v",
+		snapshotID, volumeID, snapshotSize*common.MbInBytes, snapshotCreateTimeInProto,
+		*snapshotCreateTime, createSnapshotResponse)
+	return createSnapshotResponse, nil
 }
 
 func (c *controller) DeleteSnapshot(ctx context.Context, req *csi.DeleteSnapshotRequest) (
 	*csi.DeleteSnapshotResponse, error) {
 	ctx = logger.NewContextWithLogger(ctx)
 	log := logger.GetLogger(ctx)
 	log.Infof("DeleteSnapshot: called with args %+v", *req)
-	return nil, logger.LogNewErrorCode(log, codes.Unimplemented, "deleteSnapshot")
+
+	isBlockVolumeSnapshotEnabled := commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.BlockVolumeSnapshot)
+	if !isBlockVolumeSnapshotEnabled {
+		return nil, logger.LogNewErrorCode(log, codes.Unimplemented, "deleteSnapshot")
+	}
+
+	// Validate arguments
+	csiSnapshotID := req.GetSnapshotId()
```

Comment:
Add a comment before "func DeleteSnapshot" to explain DeleteSnapshotRequest contains a SnapshotId which is in this format.  This is needed for CNS delete snapshot spec.